rgdps/usecases/messages.py

```diff
@@ -0,0 +1,269 @@
+
+    recipient = await repositories.user.from_id(ctx, recipient_user_id)
+    if recipient is None:
+        return ServiceError.MESSAGES_INVALID_RECIPIENT
+
+    friendship_check = await repositories.user_relationship.check_relationship_exists(
+        ctx,
+        user_id=sender_user_id,
+        target_user_id=recipient_user_id,
+        relationship_type=UserRelationshipType.FRIEND,
+    )
+
+    block_check = await repositories.user_relationship.check_relationship_exists(
+        ctx,
+        user_id=recipient_user_id,
+        target_user_id=sender_user_id,
+        relationship_type=UserRelationshipType.BLOCKED,
+    )
+
+    if block_check:
+        return ServiceError.MESSAGES_SENDER_BLOCKED
+
+    if recipient.message_privacy is UserPrivacySetting.PRIVATE:
+        return ServiceError.MESSAGES_MESSAGES_DISABLED
+
+    if recipient.message_privacy is UserPrivacySetting.FRIENDS and not friendship_check:
+        return ServiceError.MESSAGES_SENDER_NOT_FRIENDS
+
+    message = await repositories.message.create(
+        ctx,
+        sender_user_id=sender_user_id,
+        recipient_user_id=recipient_user_id,
+        subject=subject,
+        content=content,
+    )
+
+    return message
+
+
+async def mark_message_as_seen(
+    ctx: Context,
+    user: User,
+    message_id: int,
+) -> Message | ServiceError:
+    message = await repositories.message.from_id(
+        ctx,
+        message_id=message_id,
+        include_deleted=False,
+    )
+
+    if message is None:
+        return ServiceError.MESSAGES_NOT_FOUND
+
+    if message.recipient_user_id != user.id:
+        return ServiceError.MESSAGES_INVALID_RECIPIENT
+
+    message = await repositories.message.update_partial(
+        ctx,
+        message_id=message_id,
+        seen_ts=datetime.now(),
+    )
+
+    if message is None:
+        return ServiceError.MESSAGES_NOT_FOUND
+
+    return message
+
+
+async def delete_by_user(
+    ctx: Context,
+    user: User,
+    message_id: int,
+) -> Message | ServiceError:
+    message = await repositories.message.from_id(
+        ctx,
+        message_id=message_id,
+        include_deleted=False,
+    )
+
+    if message is None:
+        return ServiceError.MESSAGES_NOT_FOUND
+
+    if message.sender_user_id != user.id and message.recipient_user_id != user.id:
+        return ServiceError.MESSAGES_INVALID_OWNER
+
+    # Turns out that geometry dash delete message just for one person.
+    sender_deleted = False
+    recipient_deleted = False
+
+    if message.sender_user_id == user.id:
+        sender_deleted = True
+    else:
+        recipient_deleted = True
+
+    message = await repositories.message.update_partial(
+        ctx,
+        message_id=message_id,
+        sender_deleted=sender_deleted,
+        recipient_deleted=recipient_deleted,
+    )
+
+    if message is None:
+        return ServiceError.MESSAGES_NOT_FOUND
+
+    return message
+
+
+async def delete(
+    ctx: Context,
+    user: User,
```

Comment:
I would probably just pass user_id here.

Reply:
resolved in 13c7f80